### PR TITLE
Course Settings Editor

### PR DESCRIPTION
**Pull request by mylesw27**
This PR intends to make the form in `instructorCourseAdminSettings` editable. This will allows users with edit permissions, to edit course settings from this page. In this PR we will allow editing of three course settings fields: `short name`, `title` and `timezone`.

Screenshot 2024-03-11 at 3 20 35 PM

### CODE REVIEW

File: apps/prairielearn/src/lib/editors.js [hunk 1]
```diff
@@ -1386,30 +1386,21 @@ export class FileModifyEditor extends Editor {
 export class CourseInfoEditor extends Editor {
   constructor(params) {
     super(params);
-    this.description = `Create infoCourse.json`;
+    this.description = `Write infoCourse.json`;
+    this.infoJson = params.infoJson;
+    this.flag = params.flag || '';
```
Comment (mylesw27) [inline apps/prairielearn/src/lib/editors.js:1391]: Instead of using a new subclass, I reused the `CourseInfoEditor` subclass. We then differentiate if we are adding a new `infoCourse.json` or reading the current file and overwriting the changes in the controller. Those settings are then passed through the params here.

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts [hunk 2]
```diff
@@ -115,30 +150,44 @@ export function InstructorCourseAdminSettings({
...
+                        ? resLocals.authz_data.has_course_permission_edit &&
+                          !resLocals.course.example_course
+                          ? html`
+                              <button
+                                id="save-button"
+                                type="submit"
+                                class="btn btn-primary mb-2"
+                                name="__action"
+                                value="update_configuration"
+                              >
```
Comment (mylesw27) [inline apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts:165]: Is there precedent for enabling the button only after changes have been made in the form? There was mention of this in the project description but I am not sure how to accomplish this. I can start writing a new script for it but thought I would check first.

Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts:167]: Follow up: The only similar element I found was in the `fileEditor` which did not behave exactly as we wanted here. I wrote a script in `/assets/scripts/saveButtonEnabling.ts` that checks the input fields for differences and then enables the save button if any fields have changed.

File: apps/prairielearn/assets/scripts/saveButtonEnabling.ts
Comment (mylesw27) [inline apps/prairielearn/assets/scripts/saveButtonEnabling.ts:1]: I was not able find Bootstrap behavior or an existing script to enable the save button after changes have been made. I tried to leave as many comments as I could through the file to detail what is happening in each step. If there is a better way to handle behavior here, let me know.

Comment (mylesw27) [reply apps/prairielearn/assets/scripts/saveButtonEnabling.ts:1]: Discussed save button enabling with @nwalters512 on Zoom and refactored this script to make it more universal for use on other pages moving forward.

File: apps/prairielearn/assets/scripts/lib/saveButtonEnabling.ts [hunk 3]
```diff
@@ -1,25 +1,21 @@
...
+export function saveButtonEnabling(form: HTMLFormElement, saveButton: HTMLButtonElement) {
   saveButton.setAttribute('disabled', 'true');
 
   // Create a record to track if each value has changed. We are not storing initial values, but rather a boolean for each input to track if the value has changed. This is important so we don't have to read all values for each input on every change event when we check if there are differences.
   const valueHasChanged: Record<string, boolean> = {};
-  $('input, select').each(function () {
-    valueHasChanged[this.id] = false;
+  form.querySelectorAll('input, select').forEach((element) => {
+    valueHasChanged[element.id] = false;
   });
```
Comment (jonatanschroeder) [inline apps/prairielearn/assets/scripts/lib/saveButtonEnabling.ts:8]: Please check this for a simpler method: https://github.com/PrairieLearn/PrairieLearn/blob/6a1af4a93c80eebcc4348b49ba8389adac152abf/apps/prairielearn/assets/scripts/lib/confirmOnUnload.ts#L1-L10

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts [hunk 4]
```diff
@@ -18,37 +19,79 @@ router.get(
...
+      );
+    }
+
+    if (req.body.__action === 'update_configuration') {
+      const courseInfo = JSON.parse(
+        await fs.readFileSync(path.join(res.locals.course.path, 'infoCourse.json'), 'utf8'),
+      );
+      courseInfo.name = req.body.short_name;
+      courseInfo.title = req.body.title;
+      courseInfo.timezone = req.body.display_timezone;
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts:50]: I think we need to be a little careful handling this column. Currently some courses are relying on the default that's set for this value during syncing, and might also rely on inherited values from the institution. I think we want to do the following:

- If `timezone` is specified in the file and is a string, write the POSTed timezone back to the file.
- If `timezone` is not present in the file or is not a string, only write the timezone into the file if it's different from `res.locals.course.display_timezone`.

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts [hunk 5]
```diff
@@ -74,14 +94,31 @@ export function InstructorCourseAdminSettings({
...
+                    class="form-control" 
+                    id="display_timezone" 
+                    name="display_timezone" 
+                    ${courseInfoExists && resLocals.authz_data.has_course_permission_edit ? '' : 'disabled'}
+                  >
+                    ${availableTimezones.map(
+                      (tz, i) => html`
+                        <option
+                          value="${tz.name}"
+                          id="timezone-${i}"
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts:107]: Do these need IDs?

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts (hunk 4 above)
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts:50]: Actually, I take it back! I think it makes sense to always explicitly write the timezone back to the file.

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts [hunk 6]
```diff
@@ -18,37 +19,79 @@ router.get(
...
+  asyncHandler(async (req, res, next) => {
+    if (!res.locals.authz_data.has_course_permission_edit || res.locals.course.example_course) {
+      return next(
+        error.make(403, 'Access denied (must be course editor and must not be example course)'),
+      );
+    }
+
+    if (req.body.__action === 'update_configuration') {
+      const courseInfo = JSON.parse(
+        await fs.readFileSync(path.join(res.locals.course.path, 'infoCourse.json'), 'utf8'),
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts:46]: Commenting before I forget: one should basically never use `*Sync` variants of `fs` functions in production code, since they'll block the event loop.

File: apps/prairielearn/src/lib/editors.js (hunk 1 above)
Comment (mylesw27) [reply apps/prairielearn/src/lib/editors.js]: Sorry about that! Now using the `FileModifyEditor` in [79f4f05](https://github.com/PrairieLearn/PrairieLearn/pull/9581/commits/79f4f05b8d78db39a94b29e232a22bdc4c552b34).

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts [hunk 7]
```diff
@@ -37,7 +54,9 @@ export function InstructorCourseAdminSettings({
                     id="short_name"
                     name="short_name"
                     value="${resLocals.course.short_name}"
-                    disabled
+                    pattern="[a-zA-Z0-9_ ]+"
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts]: No, was trying to implement something similar to the  `courseRequest` page but it not necessary. Removed in [79f4f05](https://github.com/PrairieLearn/PrairieLearn/pull/9581/commits/79f4f05b8d78db39a94b29e232a22bdc4c552b34).

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts [hunk 8]
```diff
@@ -74,14 +94,31 @@ export function InstructorCourseAdminSettings({
...
+                          ${tz.name === resLocals.course.display_timezone ? 'selected' : ''}
+                        >
+                          (UTC
+                          ${`${tz.utc_offset.hours ? tz.utc_offset.hours : '00'}:${
+                            tz.utc_offset.minutes
+                              ? tz.utc_offset.minutes > 0
+                                ? tz.utc_offset.minutes
+                                : tz.utc_offset.minutes * -1
+                              : '00'
+                          }) ${tz.name}`}
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts]: Added a `formatTimezone` function in `lib/timezones.ts` and using it in the html file in [79f4f05](https://github.com/PrairieLearn/PrairieLearn/pull/9581/commits/79f4f05b8d78db39a94b29e232a22bdc4c552b34).

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts [hunk 9]
```diff
@@ -18,37 +19,79 @@ router.get(
...
-          res.redirect(req.originalUrl);
-        }
+router.post(
+  '/',
+  asyncHandler(async (req, res, next) => {
+    if (!res.locals.authz_data.has_course_permission_edit || res.locals.course.example_course) {
+      return next(
+        error.make(403, 'Access denied (must be course editor and must not be example course)'),
+      );
+    }
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts]: These are now separate conditionals that throw errors in [79f4f05](https://github.com/PrairieLearn/PrairieLearn/pull/9581/commits/79f4f05b8d78db39a94b29e232a22bdc4c552b34).

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts [hunk 10]
```diff
@@ -18,37 +19,79 @@ router.get(
...
-});
+      const serverJob = await editor.prepareServerJob();
+      try {
+        await editor.executeWithServerJob(serverJob);
+        return res.redirect(req.originalUrl);
+      } catch (err) {
+        return res.redirect(res.locals.urlPrefix + '/edit_error/' + serverJob.jobSequenceId);
+      }
+    } else {
+      next(error.make(400, `unknown __action: ${req.body.__action}`));
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts]: Throwing the error in [79f4f05](https://github.com/PrairieLearn/PrairieLearn/pull/9581/commits/79f4f05b8d78db39a94b29e232a22bdc4c552b34).

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts (hunk 4 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts]: Always explicitly writing the timezone back to the file. Also, as of [79f4f05](https://github.com/PrairieLearn/PrairieLearn/pull/9581/commits/79f4f05b8d78db39a94b29e232a22bdc4c552b34) we are now writing the default institution timezone when creating the file.

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts (hunk 5 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts]: No. Removed in [79f4f05](https://github.com/PrairieLearn/PrairieLearn/pull/9581/commits/79f4f05b8d78db39a94b29e232a22bdc4c552b34).

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts (hunk 6 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts]: Updated  all uses of `readFileSync` to `readFile` in [79f4f05](https://github.com/PrairieLearn/PrairieLearn/pull/9581/commits/79f4f05b8d78db39a94b29e232a22bdc4c552b34). Also, thanks for the reading on these and learning about the event loop in Node!

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts [hunk 11]
```diff
@@ -115,30 +152,52 @@ export function InstructorCourseAdminSettings({
...
-                              View course configuration
-                            </a>
-                            in <code>infoCourse.json</code>
-                          `
+                  ${
+                    coursePathExists && courseInfoExists
+                      ? resLocals.authz_data.has_course_permission_view
+                        ? resLocals.authz_data.has_course_permission_edit &&
+                          !resLocals.course.example_course
+                          ? html`
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts]: This is now a new component function called `displayEditButtons` that utilizes early returns.

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts]: Sweet! Want to update the other places we have this logic too?

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts [hunk 12]
```diff
@@ -18,37 +19,79 @@ router.get(
...
+      } catch (err) {
+        return res.redirect(res.locals.urlPrefix + '/edit_error/' + serverJob.jobSequenceId);
+      }
+    }
+
+    if (req.body.__action === 'add_configuration') {
+      const infoJson = {
+        uuid: uuidv4(),
+        name: path.basename(res.locals.course.path),
+        title: path.basename(res.locals.course.path),
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.ts:101]: I can't find where in that commit this happened; AFAICT you actually removed the entire JSON that we're constructing the file with?

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts]: Updated to use this in `administratorInstitutions` and `institutionAdminGeneral` in [feb47b1](https://github.com/PrairieLearn/PrairieLearn/pull/9581/commits/feb47b166ccac4219905c3a6ca39f5067df000fd).

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts]: Updated name to `EditActions` in [e9b0156](https://github.com/PrairieLearn/PrairieLearn/pull/9581/commits/e9b0156781e1b782b6b810d587d347e89e0fc9a7).

File: apps/prairielearn/src/lib/editors.js [hunk 13]
```diff
@@ -1386,30 +1386,20 @@ export class FileModifyEditor extends Editor {
 export class CourseInfoEditor extends Editor {
   constructor(params) {
     super(params);
-    this.description = `Create infoCourse.json`;
+    this.description = `Write infoCourse.json`;
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/editors.js:1389]: I think we actually want to revert this change and the `commitMessage` change below, since we're truly only using this to create new files. Maybe name this `CourseInfoCreateEditor` to make it clear from the name?

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts [hunk 14]
```diff
@@ -74,14 +102,27 @@ export function InstructorCourseAdminSettings({
                 </div>
                 <div class="form-group">
                   <label for="display_timezone">Timezone</label>
-                  <input
-                    type="text"
+                  <select
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts:105]: Nit: maybe move this above the "Institution" part so that we can separate the editable fields from the non-editable ones?

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts [hunk 15]
```diff
@@ -186,3 +211,63 @@ function CourseDirectoryMissingAlert({
...
+  coursePathExists: boolean;
+  courseInfoExists: boolean;
+  hasCoursePermissionView: boolean;
+  hasCoursePermissionEdit: boolean;
+  exampleCourse: boolean;
+  urlPrefix: string;
+  navPage: string;
+}) {
+  if (!coursePathExists || !courseInfoExists || !hasCoursePermissionView) {
+    return;
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts:233]: Nit, we'll soon be deprecating the usage of `null`/`undefined` in templates.

```suggestion
    return '';
```

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts [hunk 16]
```diff
@@ -186,3 +211,63 @@ function CourseDirectoryMissingAlert({
...
+        <a
+          data-testid="edit-course-configuration-link"
+          href="${urlPrefix}/${navPage}/file_edit/infoCourse.json"
+        >
+          Edit course configuration
+        </a>
+        in <code>infoCourse.json</code>
+      </p>
+    `;
+  }
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts:272]: Nit, it's nice to reduce indentation (this suggestion will need Prettier's help):

```suggestion
  }
  
  return html`
      <button
        id="save-button"
        type="submit"
        class="btn btn-primary mb-2"
        name="__action"
        value="update_configuration"
      >
        Save
      </button>
      <button
        id="cancel-button"
        type="button"
        class="btn btn-secondary mb-2"
        onclick="window.location.reload()"
      >
        Cancel
      </button>
      <p class="mb-0">
        <a
          data-testid="edit-course-configuration-link"
          href="${urlPrefix}/${navPage}/file_edit/infoCourse.json"
        >
          Edit course configuration
        </a>
        in <code>infoCourse.json</code>
      </p>
  `;
```

**Review by nwalters512 — COMMENTED** [review]
A few minor nits, but beyond those, I think this is good to go! 🎉

File: apps/prairielearn/src/lib/editors.js (hunk 13 above)
Comment (mylesw27) [reply apps/prairielearn/src/lib/editors.js]: Reverted changes here and changed name to `CourseInfoCreateEditor` in [1f9d924](https://github.com/PrairieLearn/PrairieLearn/pull/9581/commits/1f9d9242b1ab757397ce2a399f68924f58891b8e).

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts (hunk 14 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts:91]: Moved the timezone field above the institution field in [1f9d924](https://github.com/PrairieLearn/PrairieLearn/pull/9581/commits/1f9d9242b1ab757397ce2a399f68924f58891b8e).

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts (hunk 15 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts]: Updated locally in [1f9d924](https://github.com/PrairieLearn/PrairieLearn/pull/9581/commits/1f9d9242b1ab757397ce2a399f68924f58891b8e).

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts (hunk 16 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts]: Updated locally in [1f9d924](https://github.com/PrairieLearn/PrairieLearn/pull/9581/commits/1f9d9242b1ab757397ce2a399f68924f58891b8e).

File: apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts [hunk 17]
```diff
@@ -186,3 +211,62 @@ function CourseDirectoryMissingAlert({
...
+
+  if (!hasCoursePermissionEdit || exampleCourse) {
+    return html`
+      <p class="mb-0">
+        <a href="${urlPrefix}/${navPage}/file_view/infoCourse.json"> View course configuration </a>
+        in <code>infoCourse.json</code>
+      </p>
+    `;
+  }
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminSettings/instructorCourseAdminSettings.html.ts:243]: Entirely a nit:

```suggestion
  }

```

**Review by nwalters512 — APPROVED** [review]
